api/client.js

```diff
@@ -120,56 +120,62 @@ function connect(url, options={}, callback=null) {
-    See the docstring for the Connection class for information on how to use
-    the connection instance. Redirection errors are automatically handled by
-    this function, so any error is a real connection problem.
+    callback receives an error, or a {conn, logout} result including a
+    connection object and a logout function that can be used to close the
+    connection. If there are no errors, the connection can be used to
+    send/receive messages to and from the Juju controller or model, and to get
+    access to the available facades (through conn.facades). See the docstring
+    for the Connection class for information on how to use the connection
+    instance. Redirection errors are automatically handled by this function, so
+    any error is a real connection problem.
+  @return {Promise} This promise will be rejected if there is an error
+    connecting, or resolved with a new {conn, logout} object. Note that the
+    promise will not be resolved or rejected if a callback is provided.
 */
 function connectAndLogin(url, credentials, options, callback) {
-  // Connect to Juju.
-  connect(url, options, (err, juju) => {
-    if (err) {
-      callback(err, null, null);
-      return;
-    }
-    // Authenticate.
-    juju.login(credentials, (err, conn) => {
-      if (!err) {
-        callback(null, conn, juju.logout.bind(juju));
-        return;
-      }
-      if (!juju.isRedirectionError(err)) {
-        callback(err, null, null);
+  return new Promise((resolve, reject) => {
+    const handler = createAsyncHandler(callback, resolve, reject);
+    // Connect to Juju.
+    connect(url, options, (err, juju) => {
+      if (err) {
+        handler(err, null);
         return;
       }
-      // Redirect to the real model.
-      juju.logout();
-      for (let i = 0; i < err.servers.length; i++) {
-        const srv = err.servers[i];
-        // TODO(frankban): we should really try to connect to all servers and
-        // just use the first connection available, without second guessing
-        // that the public hostname is reachable.
-        if (srv.type === 'hostname' && srv.scope === 'public') {
-          // This is a public server with a dns-name, connect to it.
-          connectAndLogin(srv.url(url), credentials, options, callback);
+      // Authenticate.
+      juju.login(credentials, (err, conn) => {
+        if (!err) {
+          handler(null, {conn: conn, logout: juju.logout.bind(juju)});
           return;
         }
-      }
-      callback(
-        new Error('cannot connect to model after redirection'), null, null);
+        if (!juju.isRedirectionError(err)) {
+          handler(err, null);
+          return;
+        }
+        // Redirect to the real model.
+        juju.logout();
+        for (let i = 0; i < err.servers.length; i++) {
+          const srv = err.servers[i];
+          // TODO(frankban): we should really try to connect to all servers and
```

Comment:
Maybe create an issue for this so that we have an open task available for someone if they are interested in contributing.

Reply:
There is an issue already: https://github.com/juju/js-libjuju/issues/15